counsel.el

```diff
@@ -2572,7 +2573,29 @@ AG-PROMPT, if non-nil, is passed as `ivy-read' prompt argument."
               :unwind (lambda ()
                         (counsel-delete-process)
                         (swiper--cleanup))
-              :caller 'counsel-ag)))
+              :caller caller)))
+
+(cl-defmacro def-counsel-ag-command (name &key tool-name base-command-variable-name)
+  (declare (indent 1))
+  (or (boundp base-command-variable-name) (error "Variable %S not defined" base-command-variable-name))
+  `(progn
+     (autoload ',name "counsel" "" t)
+     (defun ,name (&optional initial-input initial-directory extra-args prompt)
+       ,(format
+         "Grep for a string in the current directory using %s.
+INITIAL-INPUT can be given as the initial minibuffer input.
+INITIAL-DIRECTORY, if non-nil, is used as the root directory for search.
+EXTRA-ARGS string, if non-nil, is appended to `%s'.
+PROMPT, if non-nil, is passed as `ivy-read' prompt argument."
+         tool-name
+         base-command-variable-name)
+       (interactive)
+       (let ((counsel-ag-base-command ,base-command-variable-name))
+         (counsel-ag initial-input initial-directory extra-args prompt ',name)))
```

Comment:
I really wish the other search commands had never hijacked `counsel-ag` in this way; it makes no sense to me. Instead of bloating the argument list and complicating the implementation of `counsel-ag`, the common functionality should have been extracted into separate, reusable subroutines.

This PR follows the direction of deepening the use of `counsel-ag` as a foundation for all other search commands. I really wish any refactors in this department would work in the opposite direction, towards cleaner abstractions.

Reply:
Agreed in general; do you see a graceful way to improve the code quality while maintaining compatibility with existing callers?

Reply:
I don't have the time to work on a proper refactor, but the easiest way to transition away from the current argument list is by advertising a new calling convention (see [`(elisp) Declare Form`](https://www.gnu.org/software/emacs/manual/html_node/elisp/Declare-Form.html) and deprecating the old. Before this can be done, however, command-line switch handling has to be improved and made consistent across search commands.